src/utils/daos_hdlr.c

```diff
@@ -1566,11 +1598,1197 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+
+	if (file_dfs->type == POSIX) {
+		num_bytes_written = write(file_dfs->fd, buf, size);
+	} else if (file_dfs->type == DAOS) {
+		num_bytes_written = write_dfs(file_dfs, file, buf, size);
+	} else {
+		fprintf(stderr, "File type not known: %s type=%d\n",
+			file, file_dfs->type);
+	}
+	if (num_bytes_written < 0) {
+		fprintf(stderr, "write error on %s type=%d\n",
+			file, file_dfs->type);
+	}
+	return num_bytes_written;
+}
+
+static int 
+open_dfs(struct file_dfs *file_dfs, char *file, int flags, mode_t mode)
+{
+
+	int 	  rc		= 0; 
+	dfs_obj_t *parent	= NULL;
+	char 	  *name		= NULL;
+	char 	  *dir_name	= NULL;
+
+	parse_filename_dfs(file, &name, &dir_name);
+	assert(dir_name);
+	rc = dfs_lookup(file_dfs->dfs, dir_name, O_RDWR, &parent, NULL, NULL);
+	if (parent == NULL) {
+		fprintf(stderr, "dfs_lookup %s failed with error %d\n", dir_name, rc);
+		D_GOTO(out, rc = 1);
+	}
+	rc = dfs_open(file_dfs->dfs, parent, name, mode | S_IFREG,
+		flags, 0, 0, NULL, &(file_dfs->obj));
+	if (rc != 0) {
+		fprintf(stderr, "dfs_open %s failed (%d)\n", name, rc);
+	}
+out:
+	if (parent != NULL) {
+		rc = dfs_release(parent);
+		if (rc != 0) {
+			fprintf(stderr, "dfs_release %s failed with error %d\n",
+				dir_name, rc);
+		}
+	}
+	if (name != NULL)
+		free(name);
+	if (dir_name != NULL)
+		free(dir_name);
+	return rc;
+}
+
+int
+file_open(struct file_dfs *file_dfs, char *file, int flags, ...)
+{
+	/* extract the mode */
+	int	rc	 = 0;
+	int	mode_set = 0;
+	mode_t	mode	 = 0;
+
+	if (flags & O_CREAT) {
+		va_list ap;
+		va_start(ap, flags);
+		mode = va_arg(ap, mode_t);
+		va_end(ap);
+		mode_set = 1;
+	}
+
+	if (file_dfs->type == POSIX) {
+		if (mode_set) {
+			file_dfs->fd = open(file, flags, mode);
+		} else {
+			file_dfs->fd = open(file, flags);
+		}
+		if (file_dfs->fd < 0) {
+			rc = EINVAL;
+			fprintf(stderr, "file_open failed on %s: %d\n", file, file_dfs->fd);
+		}
+	} else if (file_dfs->type == DAOS) {
+		rc = open_dfs(file_dfs, file, flags, mode);
+		if (rc != 0) {
+			fprintf(stderr, "file_open failed on %s: %d\n", file, rc);
+		}
+	} else {
+		rc = 1;
+		fprintf(stderr, "File type not known: %s type=%d\n",
+			file, file_dfs->type);
+	}
+	return rc;
+}
+
+static int
+mkdir_dfs(struct file_dfs *file_dfs, const char *path, mode_t *mode)
+{
+	int 		rc      = 0;
+	dfs_obj_t 	*parent = NULL;
+	char 		*name   = NULL;
+	char 		*dname  = NULL;
+
+	parse_filename_dfs(path, &name, &dname);
+	if (dname == NULL || name == NULL) {
+		fprintf(stderr, "parsing filename failed, %s\n", dname);
+		D_GOTO(out, rc = 1);
+	}
+	rc = dfs_lookup(file_dfs->dfs, dname, O_RDWR, &parent, NULL, NULL);
+	if (parent == NULL) {
+		fprintf(stderr, "dfs_lookup %s failed \n", dname);
+		D_GOTO(out, rc = 1);
+	}
+	rc = dfs_mkdir(file_dfs->dfs, parent, name, *mode, 0);
+	if (rc != 0) {	
+		/* continue if directory exists, fail for other errors */
+		fprintf(stderr, "dfs_mkdir %s failed, %s\n", name, strerror(rc));
+	}
+out:
+	if (parent != NULL) {
+		rc = dfs_release(parent);
```

Comment:
@daltonbohning well, not entirely sure the best way to do this, because if I just return if dfs_mkdir fails, then I don't release the parent or memory from dirname, name? I think that when dfs_mkdir runs it does log a message, so will dfs_release. I guess the function will return though with the error code from dfs_release if it fails.

Reply:
@dsikich Right, we do still need to call dfs_release if dfs_mkdir fails. But if dfs_mkdir fails, and dfs_release succeeds, then rc=0 will be returned.
In mpifileutils, I recently added something like:
```
tmp_rc = dfs_release(...)
if (tmp_rc && (rc != -1)) {
    // log error and set errno accordingly
}
```
Basically, ignore a failed dfs_release if and only if dfs_mkdir already failed.
In dfs.c, I think this is handled in some places by just simply ignoring the rc from dfs_release in situations like this. For functions like "close", etc., a failed release is definitely problem. But for "transient" cases like this, it seems that dfs.c doesn't care.
But... I'm not sure if that's intentional or not.
Maybe @mchaarawi has an opinion.

Reply:
@daltonbohning using tmp_rc for now